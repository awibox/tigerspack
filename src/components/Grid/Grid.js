import React from 'react';
import PropTypes from 'prop-types';

const Grid = (props) => {
  const {
    align,
    children,
    flow,
    stretch,
    padding,
    size,
    valign,
    wrap,
    ...otherProps
  } = props;

  const stylesObject = {};

  const posNameStab = (prop) => {
    if (prop === 'top' || prop === 'left') {
      return 'flex-start';
    }
    if (prop === 'bottom' || prop === 'right') {
      return 'flex-end';
    }
    return prop;
  };

  const isColumn = flow === 'column';

  if (align) {
    stylesObject[isColumn ? 'alignContent' : 'justifyContent'] = posNameStab(align);
    stylesObject[isColumn ? 'alignItems' : 'justifyItems'] = posNameStab(align);
  }

  if (valign) {
    stylesObject[isColumn ? 'justifyContent' : 'alignContent'] = posNameStab(valign);
    stylesObject[isColumn ? 'justifyItems' : 'alignItems'] = posNameStab(valign);
  }

  if (stretch) {
    switch (stretch) {
      case 'width': {
        stylesObject.width = '100%';
        break;
      }
      case 'height': {
        stylesObject.height = '100%';
        break;
      }
      default: {
        stylesObject.width = '100%';
        stylesObject.height = '100%';
        break;
      }
    }
  }

  if (size < 13) {
    stylesObject[isColumn ? 'minHeight' : 'minWidth'] = `${(100 / 12) * size}%`;
  }

  const gridStyle = {
    display: 'flex',
    flexFlow: flow,
    padding: `${padding}px`,
    boxSizing: 'border-box',
    flexWrap: wrap ? 'wrap' : 'nowrap',
    ...stylesObject,
  };

  return (
    <div css={gridStyle} {...otherProps}>
      {children}
    </div>
  );
};

Grid.propTypes = {
  align: PropTypes.oneOf(['left', 'right', 'center']),
  valign: PropTypes.oneOf(['top', 'bottom', 'center']),
  children: PropTypes.any.isRequired,
  flow: PropTypes.oneOf(['row', 'column']),
  padding: PropTypes.number,
  stretch: PropTypes.oneOf(['width', 'height', 'full']),
  size: PropTypes.number,
  wrap: PropTypes.bool,
};

Grid.defaultProps = {
  padding: 0,
};

export default Grid;
