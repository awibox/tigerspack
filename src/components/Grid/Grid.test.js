import React from 'react';
import { shallow } from 'enzyme';
import Grid from './Grid';


describe('Grid', () => {
  it('Default Grid', () => {
    const component = shallow(<Grid>Test Grid</Grid>);
    expect(component).toMatchSnapshot();
  });
  it('Grid align', () => {
    const component = shallow(<Grid align={'left'}>Test Grid</Grid>);
    expect(component.get(0).props.css.justifyContent).toEqual('flex-start');
    expect(component.get(0).props.css.justifyItems).toEqual('flex-start');
  });
  it('Grid align right', () => {
    const component = shallow(<Grid align={'right'}>Test Grid</Grid>);
    expect(component.get(0).props.css.justifyContent).toEqual('flex-end');
    expect(component.get(0).props.css.justifyItems).toEqual('flex-end');
  });
  it('Grid align center', () => {
    const component = shallow(<Grid align={'center'}>Test Grid</Grid>);
    expect(component.get(0).props.css.justifyContent).toEqual('center');
    expect(component.get(0).props.css.justifyItems).toEqual('center');
  });
  it('Grid valign', () => {
    const component = shallow(<Grid valign={'top'}>Test Grid</Grid>);
    expect(component.get(0).props.css.alignContent).toEqual('flex-start');
    expect(component.get(0).props.css.alignItems).toEqual('flex-start');
  });
  it('Grid align column', () => {
    const component = shallow(<Grid flow={'column'} align={'left'}>Test Grid</Grid>);
    expect(component.get(0).props.css.flexFlow).toEqual('column');
    expect(component.get(0).props.css.alignContent).toEqual('flex-start');
    expect(component.get(0).props.css.alignItems).toEqual('flex-start');
  });
  it('Grid valign column', () => {
    const component = shallow(<Grid flow={'column'} valign={'top'}>Test Grid</Grid>);
    expect(component.get(0).props.css.justifyContent).toEqual('flex-start');
    expect(component.get(0).props.css.justifyItems).toEqual('flex-start');
  });
  it('Grid stretch width', () => {
    const component = shallow(<Grid stretch={'width'}>Test Grid</Grid>);
    expect(component.get(0).props.css.width).toEqual('100%');
  });
  it('Grid stretch height', () => {
    const component = shallow(<Grid stretch={'height'}>Test Grid</Grid>);
    expect(component.get(0).props.css.height).toEqual('100%');
  });
  it('Grid stretch full', () => {
    const component = shallow(<Grid stretch={'full'}>Test Grid</Grid>);
    expect(component.get(0).props.css.width).toEqual('100%');
    expect(component.get(0).props.css.height).toEqual('100%');
  });
  it('Grid width', () => {
    const component = shallow(<Grid width={1}>Test Grid</Grid>);
    expect(component.get(0).props.css.minWidth).toEqual(`${(100 / 12)}%`);
    expect(component.get(0).props.css.maxWidth).toEqual(`${(100 / 12)}%`);
  });
  it('Grid height', () => {
    const component = shallow(<Grid height={1}>Test Grid</Grid>);
    expect(component.get(0).props.css.minHeight).toEqual(`${(100 / 12)}%`);
    expect(component.get(0).props.css.maxHeight).toEqual(`${(100 / 12)}%`);
  });
  it('Grid padding', () => {
    const component = shallow(<Grid padding={10}>Test Grid</Grid>);
    expect(component.get(0).props.css.padding).toEqual('10px');
  });
  it('Grid wrap', () => {
    const component = shallow(<Grid wrap>Test Grid</Grid>);
    expect(component.get(0).props.css.flexWrap).toEqual('wrap');
  });
});
