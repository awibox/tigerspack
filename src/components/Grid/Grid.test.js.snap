// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid Default Grid 1`] = `
<EmotionCssPropInternal
  __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Component"
  __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
  css={
    Object {
      "boxSizing": "border-box",
      "display": "flex",
      "flexFlow": undefined,
      "flexWrap": "nowrap",
      "padding": "0px",
    }
  }
>
  Test Grid
</EmotionCssPropInternal>
`;
